subiquity/server/controllers/mirror.py

```diff
@@ -50,6 +50,28 @@ class MirrorController(SubiquityController):
                     'enum': ['universe', 'multiverse', 'restricted',
                              'contrib', 'non-free']
                 }
+            },
+            "preferences": {
+                "type": "array",
+                "items": {
+                    "type": "object",
+                    "properties": {
+                        "package": {
+                            "type": "string",
+                        },
+                        "pin": {
+                            "type": "string",
```

Comment:
1. What are thoughts on the Pin value?  
If desired, we could be more specific and limit Pin to just having lines that start with origin, release, version (are there others?)
(That's a question, not a demand)

2. The apt_preferences manpage documents Pin like so
```
Pin: origin "ftp.de.debian.org"
```
Do things get weird in the yaml if part of it is quoted?

Reply:
Thanks for the feedback!

> 1. What are thoughts on the Pin value?
>    If desired, we could be more specific and limit Pin to just having lines that start with origin, release, version (are there others?)
>    (That's a question, not a demand)

I have been hesitant about that very thing. At first I wanted to establish stricter rules but I realized that the driver for this use-case is to avoid the current workaround (which consists of using `late_commands`).

When using `late_commands`, there is no safety net to guard against non-existent keywords or invalid stanzas. So I have assumed that using stricter rules would not bring that much more to the table. But I would be happy to change that if you think it would be better.

> 2. The apt_preferences manpage documents Pin like so
> 
> ```
> Pin: origin "ftp.de.debian.org"
> ```
> 
> Do things get weird in the yaml if part of it is quoted?

It should be fine, APT behaves the same if the origin field is quoted or unquoted:
https://salsa.debian.org/apt-team/apt/-/blob/main/apt-pkg/versionmatch.cc#L124
```cpp
   if (Type == Origin)
   {
      if (Data[0] == '"' && Data.length() >= 2 && Data.end()[-1] == '"')
	 OrSite = Data.substr(1, Data.length() - 2);
      else
	 OrSite = Data;
      return;
   }   
```
as for YAML itself, these values are equivalent (as far as my tests seem to show):
```yaml
pin: origin "ftp.de.debian.org"
pin: 'origin "ftp.de.debian.org"'
pin: "origin \"ftp.de.debian.org\""
```
However, I can foresee the character `"\n"` as being problematic when de-serializing the data into stanzas. E.g.:
```yaml
preferences:
  {package: dummy, pin: "some\nvalue", pin-priority: -1}
```
->
```
Package: dummy
Pin: some
value
Pin-Prioriy: -1
```
Maybe it's worth rejecting this character from pin, package and pin-priority. To be fair, there might be other characters that could make things go wrong. But since the stanzas from APT are more user-friendly than machine-friendly, it's difficult to come up with something perfect IMO.

Thanks,
Olivier